### PR TITLE
tests to prevent no row insertion in DB when there is empty/no local …

### LINKED ISSUE #346: Add tests to prevent empty/no-op local changes
> OK I'm just worried there are no failing tests. what tests should we add that should have failed as a result of this change?

The squashing algorithm works fine with empty changes. There's no API besides checking the table for row count change. Not sure how to test this reliably. I'll think a bit more about this.

_Originally posted by @deepankarb in https://github.com/google/android-fhir/issues/303#issuecomment-795444119_

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
…changes

**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #346

**Description**
Test case added for no row should be inserted in database if there is update with empty/ no local changes

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 1]
```diff
@@ -134,6 +134,26 @@ class DatabaseImplTest {
...
+    val patient: Patient = testingUtils.readFromFile(Patient::class.java, "/date_test_patient.json")
+    database.insert(patient)
+    patient.gender = Enumerations.AdministrativeGender.FEMALE
+    database.update(patient)
+    patient.name[0].family = "TestPatient"
+    database.update(patient)
+
+    var thirdlocalChangeCount = 0
+    database.getAllLocalChanges().map { thirdlocalChangeCount = it.token.ids.size }
+    assertThat(thirdlocalChangeCount).isEqualTo(3)
```
[the remark this answers is not in the document]
Comment (PallaviGanorkar) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:148]: done

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 2]
```diff
@@ -134,6 +134,26 @@ class DatabaseImplTest {
...
+
+    // update patient with no local change
+    database.update(patient)
+    var fourthLocalChangeCount = 0
+    database.getAllLocalChanges().map { fourthLocalChangeCount = it.token.ids.size }
+    assertThat(fourthLocalChangeCount).isEqualTo(3)
```
[the remark this answers is not in the document]
Comment (PallaviGanorkar) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:154]: done

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 3]
```diff
@@ -134,6 +134,38 @@ class DatabaseImplTest {
...
+  @Test
+  fun update_existentResourceWithNoChange_shouldNotUpdateResource() = runBlocking {
+    val patient: Patient = testingUtils.readFromFile(Patient::class.java, "/date_test_patient.json")
+    database.insert(patient)
+    patient.gender = Enumerations.AdministrativeGender.FEMALE
+    database.update(patient)
+    patient.name[0].family = "TestPatient"
+    database.update(patient)
+    val patientString = services.parser.encodeResourceToString(patient)
+    val squashed3localChanges =
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:146]:
```suggestion
    val squashedLocalChange =
```

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 4]
```diff
@@ -134,6 +134,38 @@ class DatabaseImplTest {
...
+      database.getAllLocalChanges().single { it.localChange.resourceId.equals(patient.logicalId) }
+    assertThat(squashed3localChanges.token.ids.size).isEqualTo(3)
+    with(squashed3localChanges.localChange) {
+      assertThat(resourceId).isEqualTo(patient.logicalId)
+      assertThat(resourceType).isEqualTo(patient.resourceType.name)
+      assertThat(type)
+        .isEqualTo(com.google.android.fhir.db.impl.entities.LocalChangeEntity.Type.INSERT)
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:153]: Question: Is your AndroidStudio configured to show these suggestions?

```suggestion
        .isEqualTo(LocalChangeEntity.Type.INSERT)
```

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 5]
```diff
@@ -134,6 +134,38 @@ class DatabaseImplTest {
...
+      assertThat(payload).isEqualTo(patientString)
+    }
+    // update patient with no local change
+    database.update(patient)
+    val squashedlocalChangeWithNoUpdate =
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:158]:
```suggestion
    val squashedLocalChangeWithNoFurtherUpdate =
```

**Review by jingtang10 — APPROVED** [review]
Approved - but please address the comments first

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt (hunk 3 above)
Comment (PallaviGanorkar) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:146]: Done

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt (hunk 4 above)
Comment (PallaviGanorkar) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:153]: Done

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt (hunk 5 above)
Comment (PallaviGanorkar) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:158]: Done
